### PR TITLE
Allow basic editing for supported shapes and display rotation for rectangle and ellipse

### PR DESCRIPTION
**Pull request by pllim**
### Description

This pull request is to allow basic shape editing. Also shows rotation angle for some shapes.

Screenshot 2022-08-05 223316

Related:

* https://github.com/glue-viz/glue/pull/2235
* https://github.com/spacetelescope/jdaviz/pull/729

Address the centering part of #634 because rotation is blocked by https://github.com/glue-viz/glue-astronomy/issues/74 .

### TODO

- [x] Allow edit of region center
- [x] Allow edit of region radius/width
- [x] Allow edit of angles. While edit is possible, we cannot get that rotated shape back out. This needs upstream fix, so disable for now.
- [x] What about composite regions? Edit is disabled for composite regions.
- [x] What about masked subset? Edit is disabled for masked subset.
- [x] Have clear message for weird shapes where edits not allowed. Not needed, we just hide the GUI.
- [x] Make sure this works with other stuff. ~For example, does Aperture Photometry take rotated Subset?~
- [x] Write tests
- [x] Update docs
- [ ] Optional: Why does this not play nice with `ipypopout`?

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [ ] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [x] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Did the CI pass? If not, are the failures related?
- [x] Is a change log needed? If yes, is it added to `CHANGES.rst`?
- [x] Is a milestone set?
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)? [🐱](https://jira.stsci.edu/browse/JDAT-1959)

### CODE REVIEW

**Review by duytnguyendtn — COMMENTED** [review]
Currently working on a review, but wanted to get this feature request in, in the mean time: could it be possible to apply the update when I press "Enter" on a field? Even better if it automatically moves me down to the next field? 🙏

**Comment by pllim** [comment]
> Even better if it automatically moves me down to the next field?

That is some front-end dark magic that I don't know how to do. If you do, please suggest the changes!

Does `<tab>` already works? I never tried it.

File: jdaviz/configs/default/plugins/subset_plugin/subset_plugin.vue
Comment (duytnguyendtn) [inline jdaviz/configs/default/plugins/subset_plugin/subset_plugin.vue:34]: I **think** it would be something like this:

```suggestion
          <v-text-field
            :label="item.name"
            v-model.number="item.value"
            @keyup.enter="update_subset"
            type="number"
            :disabled="item.name=='Angle'"
          ></v-text-field>
```

Injecting it in on myside seems to do the trick

**Comment by duytnguyendtn** [comment]
> Does <tab> already works? 

Come to think about it, maybe <tab> is sufficient if my cursor stayed selected on the field. If the former suggestion works (press enter to auto update), and my cursor stayed on the same field, then I could just press tab to advance. The problem currently is when I press enter, it takes my cursor off, so I have to go back to my mouse to reselect on the next field

File: jdaviz/configs/default/plugins/subset_plugin/subset_plugin.vue
Comment (pllim) [reply jdaviz/configs/default/plugins/subset_plugin/subset_plugin.vue:34]: Hmm... but this will:

(a) make this plugin behavior different from the other ones; and
(b) might prematurely update a shape when someone accidentally hits enter but really want to also update other fields?

**Comment by pllim** [comment]
> when I press enter, it takes my cursor off

But why are you pressing Enter in the first place? It currently does not do anything. I think this applies to all the plugins.

**Comment by duytnguyendtn** [comment]
> It currently does not do anything

Sorry, that comment was meant for if we go with the "press enter to update" direction

File: jdaviz/configs/default/plugins/subset_plugin/subset_plugin.vue
Comment (duytnguyendtn) [reply jdaviz/configs/default/plugins/subset_plugin/subset_plugin.vue:34]: > (a) make this plugin behavior different from the other ones

I'm not sure if there's a consistent pattern here. Take, for instance, the `Line Width` field for plot options: when I update the text field, it _immediately_ updates the line width, doesn't even wait for the user. I'd argue that it might be annoying to the user to have the field update while the user is trying to type the bounds, hence why I proposed `Enter` to accept. Though now that I say it, is the `UPDATE` button even necessary in this context? Why not just have it accept automatically? Which of course leads to...

> want to also update other fields?

Is there a problem with updating one field at a time? I don't necessarily see it as an issue if I update the `xcenter` and THEN the `ycenter` rather than changing both fields simultaneously

Comment (duytnguyendtn) [reply jdaviz/configs/default/plugins/subset_plugin/subset_plugin.vue:34]: Come to think about it, I wonder if the disconnect here has to do with perspective. For `Line Width` I think of that field as an actual representation of the value, hence when I change the number, I expect it to immediately update. For `Model Fit`, it's more like a command, so I fill out all the necessary boxes and then tell it to activate via the button click.

As I tested it now, it appears to be more of the "command" like paradigm, where the user fills out all the information and then actively requests the plugin to "UPDATE" the subset by pressing the button. I suppose my request is really wondering whether it would be better to change this to the "representation" paradigm to have it behave more like plot options, like `Line Width` for instance.

Comment (pllim) [reply jdaviz/configs/default/plugins/subset_plugin/subset_plugin.vue:34]: Subset update may be expensive, because it might affect other calculations etc. You would only want to update it when you are sure you want to update it.

Comment (duytnguyendtn) [reply jdaviz/configs/default/plugins/subset_plugin/subset_plugin.vue:34]: > Subset update may be expensive

I'm trying to think of cascading events that would happen on a SubsetUpdate message. Most of the heavy ones are still locked behind a button press (like Aperture Photometry and Model Fitting). The only ones I can think of are the Cubeviz spectrum collapse and line analysis, and the latter is only if the plugin is open. I feel like "Enter to Accept" is a QoL feature I'd be happy to accept in exchange. Are there other cases I'm not appreciating?

Comment (pllim) [reply jdaviz/configs/default/plugins/subset_plugin/subset_plugin.vue:34]: I feel like this needs broader discussion and is out of scope here. If we go with this, it will apply to other plugins too anyway.

File: jdaviz/configs/default/plugins/subset_plugin/subset_plugin.vue
Comment (duytnguyendtn) [inline jdaviz/configs/default/plugins/subset_plugin/subset_plugin.vue:29]: Not sure whether we want to include this in scope, but I feel like this spacing feels awkwardly large to me. Thoughts @Jenneh?
image

**Review by duytnguyendtn — COMMENTED** [review]
Ok here's my official review! Functionality and code looks good! Thanks Pey-Lian!

I have one comment concerning the spacing of the elements, and another in the above chain concerning the `UPDATE` button. IMHO, I feel these the subset tools plugin should behave more like an "inspector", similar to Plot Options. By that I mean modifications to the values should _immediately_ (or if not immediately, then at least when the user "exits" the scope of the text field by clicking off or pressing `ENTER`. Therefore, the `UPDATE` button would be largely unnecessary, like how we don't have one in Plot Options. I'll try to remember to bring this up at parking lot tomorrow for the team to discuss

File: jdaviz/configs/default/plugins/subset_plugin/subset_plugin.vue
Comment (pllim) [reply jdaviz/configs/default/plugins/subset_plugin/subset_plugin.vue:29]: I'll be happy to reduce the spacing if you tell me how to do it.

I am still hesitant about the "update immediately" model. I feel like the performance hit is going to come back and bite us as we build more features. But we can discuss tomorrow.

File: jdaviz/configs/default/plugins/subset_plugin/subset_plugin.vue
Comment (kecnry) [reply jdaviz/configs/default/plugins/subset_plugin/subset_plugin.vue:34]: I definitely think we need some standards defined here as we already use a few patterns throughout the plugins.  But in general, I'd agree that they _should_ all fit into either an action button or live-updating (but never both)... and the decision between which to use are the points brought up by both of you: (a) whether live-feedback between changing fields is useful and (b) whether the action is too expensive to update in real-time.  The conflict we currently have is that some fields that are live-updating update on keystroke and others when the input element loses focus - but that is definitely out of scope for this PR (but I'd argue deciding whether this PR should be live-updating when losing focus might not be).

File: jdaviz/configs/default/plugins/subset_plugin/subset_plugin.vue
Comment (kecnry) [reply jdaviz/configs/default/plugins/subset_plugin/subset_plugin.vue:29]: The extra spacing is caused by the nested `v-col`s inside `v-row`s.  When there is only a single column, its easiest to just remove `v-col` entirely.  Otherwise you'll want to add `class="row-no-outside-padding"` to the `v-row`.  See the third bullet in the [plugin style guide](https://jdaviz.readthedocs.io/en/latest/dev/ui_style_guide.html) (and feel free to clean that up if any of it seems confusing)

File: jdaviz/configs/default/plugins/subset_plugin/subset_plugin.vue
Comment (duytnguyendtn) [reply jdaviz/configs/default/plugins/subset_plugin/subset_plugin.vue:34]: > deciding whether this PR should be live-updating when losing focus might not be (out of scope)

You explained it MUCH more eloquently than I; thanks Kyle!

Comment (kecnry) [reply jdaviz/configs/default/plugins/subset_plugin/subset_plugin.vue:34]: here are the options for updating a traitlet from a `v-text-field` with `type="number"` that are used elsewhere already:
* live updating on keypress: use `v-model`  (see slice plugin for a good example) and then `@observe` in python.  We use this quite often and it mimics what glue widgets do, so unless there are performance limitations, this might be worth trying first.
* update on change (loss of focus) or mousepress on the up/down buttons: use `@change` and `@mousedown` to call a JS or python method to update the traitlet manually (play_pause widget does something similar - but this also uses `v-model` and just triggers acting on it at those times, you could do the same to always have the traitlet update but only trigger the “update” when this happens, or if you want to have an `@observe` still in python, then you might need extra traitlets).

File: jdaviz/configs/default/plugins/subset_plugin/subset_plugin.vue
Comment (pllim) [reply jdaviz/configs/default/plugins/subset_plugin/subset_plugin.vue:29]: I removed the `v-col` (didn't make much difference) and neither did `class="row-no-outside-padding"`, but the `class="pt-0 pb-0 mt-0 mb-0"` seems to work. I copy-pasted that from other parts of `subset_plugin.vue`.

File: jdaviz/configs/default/plugins/subset_plugin/subset_plugin.vue
Comment (pllim) [reply jdaviz/configs/default/plugins/subset_plugin/subset_plugin.vue:34]: Re: lag -- I tried with 6 4k x 2k and Imviz is already super laggy (the lag from Subset update doesn't matter much at that point because the whole thing is unusable, which is out of scope here). With 4 of those images, Imviz lag is tolerable but you can really see the lag in Subset shape catching up with each click to Update. So, I am still not convinced live-updating is a good idea.

You can try for yourselves like this:

1. Open up Imviz example notebook.
2. Load the example images twice (so you get 4 total) with this additional call:

```python
imviz.load_data(acs_47tuc_1, data_label='acs_47tuc_3')
imviz.load_data(acs_47tuc_2, data_label='acs_47tuc_4')
```

3. Draw a Subset.
4. Use this new edit feature and change some field of the Subset, click Update. See lag.

Comment (pllim) [reply jdaviz/configs/default/plugins/subset_plugin/subset_plugin.vue:34]: p.s. This step is also important: `imviz.link_data(link_type='wcs', wcs_use_affine=False)` (links them by WCS without fast approximation)

Comment (kecnry) [reply jdaviz/configs/default/plugins/subset_plugin/subset_plugin.vue]: Down the road I think we can either look into improving the performance or perhaps have this plugin check the number of links (and if fast approximation is enabled) and switch itself between live-updating and the update button.  But I'm ok considering that as a follow-up effort and moving forward with the PR with the update button.

Comment (pllim) [reply jdaviz/configs/default/plugins/subset_plugin/subset_plugin.vue]: Having the button come and go is even more confusing than just allowing live-update unconditionally. But I agree this can be future work.

Please approve if the PR is okay as-is or if there are more changes needed here. Thanks!

File: jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py
Comment (kecnry) [inline jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py:37]: is it possible to replace `is_editable` with checking the length of `subset_defintions` or are there cases where the logic needs to be different (single regions that will show in the plugin but are not considered editable)?

(I first checked to see if `has_subset_details` does the same thing, but that seems to never actually be used, so maybe that can be removed here as well).

File: jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py
Comment (kecnry) [inline jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py:202]: it might be nice to put this in the `SubsetSelect` class so its reusable.  There already is a `self.subset_select.selected_obj`, but that returns the _region_ not the _state_, and `selected_obj` is used across all selects, so we don't want to change that behavior.  Maybe this logic you have here should be `SubsetSelect.selected_subset_state`?

File: jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py
Comment (kecnry) [inline jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py:220]: is this always a rotation failure or should it just say failed to *update*?

File: jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py
Comment (kecnry) [inline jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py:217]: this might warrant an in-line comment... what does `_combine_data` do?

**Review by kecnry — COMMENTED** [review]
Seems to work well!  Is support for editing sub-subsets out-of-scope (it can display the sub-subsets, but doesn't show the logical operators between them)?

Is there a need for displaying both the edit box and the static read-only version of the subset parameters in cases where editing is supported?  It seems to me that showing both sections adds quite a bit of code overhead (needing to trace the current widget value in addition to the original value), that might be simplified if the read-only section is just _replaced_ with the widgets when editing is supported?  But then without live-updating, we might want to have a cancel/reset button to revert the widget values to the current state?

**Comment by camipacifici** [comment]
This looks very nice!
I tried both in Imviz and Specviz and here are my comments:
- Indeed there is a lag when not using the fast approximation when linking in Imviz. The lag is present (and annoying) even with just 2 images on my machine. Out of scope for this ticket, but it needs to be investigated.
- I would not want the live update to kick in when I start modifying the value. It would be nice to have it when I click the arrows, but it is useful really only for Imviz where a jump of "1" is ok. A jump of "1" in Specviz is completely useless because it goes to a completely different region of the spectrum (e.g. the starting value is 1.234um, I click the arrow, it goes to 2um), so I do not think I would ever use the arrows there.
- It would be nice to have the live update when tabbing to the next entry. Could this be an option?
- Said that, I do not find a big overhead in, for example, tabbing twice and then hitting return. I guess I need to try with different science cases, but a non-live update looks good to me (for now).

**Comment by pllim** [comment]
> it can display the sub-subsets, but doesn't show the logical operators between them

That is a known issue and has existed before this PR. In fact, https://jdaviz.readthedocs.io/en/latest/imviz/plugins.html#subset-tools documents this in the second paragraph.

**Comment by pllim** [comment]
> Is support for editing sub-subsets out-of-scope

I would say, yes. I have not found any use for compound subsets yet in Imviz. I guess it is more useful in Specviz but working on it would exceed the points for this ticket the PR is tied to.

I mean, when someone edits a compound subset so that the two sub-subsets now overlap, do we just create a single non-compound subset or throw error? There are extra stuff we need to ponder.

**Comment by pllim** [comment]
> Is there a need for displaying both the edit box and the static read-only version of the subset parameters in cases where editing is supported?

I originally combined them but find the display to be confusing because then I cannot be sure if the values are the ones I just entered or from the actual Subset state. When I edit a Subset, personally as a user, I find it useful to be able to see the current value and compare to what I just entered to make sure I am not trying to do something crazy.

File: jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py
Comment (pllim) [reply jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py:37]: > single regions that will show in the plugin but are not considered editable

Yes, masked subset state is not editable.

Comment (pllim) [reply jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py:37]: Or if in the future, we add support for drawing weird shapes but has no translator for it, that is not editable too with the current logic.

File: jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py
Comment (pllim) [reply jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py:220]: Oh... good catch! This started out as an angle edit (ironically). Will fix.

**Comment by kecnry** [comment]
> I originally combined them but find the display to be confusing because then I cannot be sure if the values are the ones I just entered or from the actual Subset state. When I edit a Subset, personally as a user, I find it useful to be able to see the current value and compare to what I just entered

I'd be curious what @Jenneh thinks, but we can always iterate after merge if/when everything else is ready to go.

File: jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py
Comment (pllim) [reply jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py:217]: Indeed. Took me a good 10 mins to recall why I picked this. I will add comment.

**Comment by pllim** [comment]
> Indeed there is a lag when not using the fast approximation when linking in Imviz. The lag is present (and annoying) even with just 2 images on my machine. Out of scope for this ticket, but it needs to be investigated.

That is the reason why we have fast approximation in the WCS transform. I do not recommend people turn fast approx. off but I turn it on to show you the worst case scenario with the lag.

**Comment by pllim** [comment]
> It would be nice to have the live update when tabbing to the next entry.

That was original request from @duytnguyendtn as well. But looks like we can defer this feature for now as Cami wants to try the plugin on different use cases first.

File: jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py
Comment (pllim) [reply jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py:202]: I tried this but does not work (tests broke). What am I doing wrong?

```diff
--- a/jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py
+++ b/jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py
@@ -187,19 +187,14 @@ class SubsetPlugin(TemplateMixin):
         """
         self.subset_definitions = []
         self.subset_types = []
-        subset_group = [s for s in self.app.data_collection.subset_groups if
-                        s.label == self.subset_selected][0]
-        subset_state = subset_group.subset_state

-        self._unpack_nested_subset(subset_state)
+        self._unpack_nested_subset(self.subset_select.selected_subset_state)

     def vue_update_subset(self, *args):
         if not self.is_editable:  # no-op
             return

-        subset_group = [s for s in self.app.data_collection.subset_groups if
-                        s.label == self.subset_selected][0]
-        subset_state = subset_group.subset_state
+        subset_state = self.subset_select.selected_subset_state

--- a/jdaviz/core/template_mixin.py
+++ b/jdaviz/core/template_mixin.py
@@ -722,6 +722,12 @@ class SubsetSelect(BaseSelectPluginComponent):
             if match is not None:
                 return match

+    @cached_property
+    def selected_subset_state(self):
+        subset_group = [s for s in self.app.data_collection.subset_groups if
+                        s.label == self.selected][0]
+        return subset_group.subset_state
+
     def selected_min_max(self, spectrum1d):
```

Comment (kecnry) [reply jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py:202]: You'll have to manually handle the cache... does replacing `@cached_property` with `@property` fix the tests?

Comment (pllim) [reply jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py:202]: Yes, that seems to fix it. Thanks!

File: jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py
Comment (pllim) [reply jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py:215]: Done.

**Review by kecnry — APPROVED** [review]
Still a few things to consider in future iterations (just summarizing points from multiple threads above, but I think we're ok to merge this as-is, once CI passes):
* re-consider live-updating and performance implications
* consult with @Jenneh if the read-only and edit version should both be shown
* re-implement popout button when issues with subsets failing to update are fixed
* enable support for angle rotation once supported upstream
* possible support for editing sub-subsets (see conversation above about why this might get difficult)

**Comment by pllim** [comment]
Thanks! I'll merge and can always open follow-up PRs if someone else submits a late review.

**Comment by pllim** [comment]
Follow-up issues:

- re-consider live-updating and performance implications: https://github.com/spacetelescope/jdaviz/issues/1565
- consult with `@Jenneh` if the read-only and edit version should both be shown: https://github.com/spacetelescope/jdaviz/issues/1566
- re-implement popout button when issues with subsets failing to update are fixed: https://github.com/spacetelescope/jdaviz/issues/1567
- enable support for angle rotation once supported upstream: https://github.com/spacetelescope/jdaviz/issues/634
- possible support for editing sub-subsets (see conversation above about why this might get difficult): https://github.com/spacetelescope/jdaviz/issues/1568

**Pull request merged**
